File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque


def solution(maps):
    return bfs(maps, 0, 0)


def bfs(graph, x, y):
    n = len(graph)
    m = len(graph[0])

    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    queue = deque()
    queue.append((x, y))

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or ny < 0 or nx >= n or ny >= m:
                continue
            if graph[nx][ny] == 0:
                continue
            if graph[nx][ny] == 1:
                graph[nx][ny] = graph[x][y] + 1
                queue.append((nx, ny))

    return graph[-1][-1]


n, m = map(int, input().split())

graph = []
for i in range(n):
    row = list(map(int, input().strip()))
    graph.append(row)

print(solution(graph))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <bits/stdc++.h>

using namespace std;

#define endl "\n"
#define FASTIO ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
#define FOR(i, n) for(int i = 0; i < (n); ++i)
#define INF (~0U >> 2)
#define PB push_back
#define MOD 1000000007

typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
int dy[4] = {1, 0, -1, 0};
int dx[4] = {0, 1, 0, -1};

int N, M;
vvi Graph;

int BFS()
{
    vvi Visited(N, vector<int>(M, -1));

    queue<pii> Queue;
    Queue.push({0, 0});
    Visited[0][0] = 1;

    while (!Queue.empty())
    {
        int x = Queue.front().first;
        int y = Queue.front().second;
        Queue.pop();

        for (int i = 0; i < 4; i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx >= 0 && nx < N && ny >= 0 && ny < M && Visited[nx][ny] == -1 && Graph[nx][ny] == 1)
            {
                Visited[nx][ny] = Visited[x][y] + 1;
                Queue.push({nx, ny});
            }
        }
    }

    return Visited[N - 1][M - 1];
}

int main()
{
    FASTIO
    cin >> N >> M;

    Graph.resize(N, vi(M, 0));

    FOR(i, N)
    {
        string input;
        cin >> input;

        FOR(j, M)
        {
            Graph[i][j] = input[j] - '0';
        }
    }

    cout << BFS() << endl;

    return 0;
}
